client/devices/client.go

```diff
@@ -0,0 +1,181 @@
+const (
+	devicesListURL = "/api/management/v2/devauth/devices"
+)
+
+type Client struct {
+	url            string
+	devicesListURL string
+	client         *http.Client
+}
+
+func NewClient(url string, skipVerify bool) *Client {
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: skipVerify},
+	}
+
+	return &Client{
+		url:            url,
+		devicesListURL: JoinURL(url, devicesListURL),
+		client: &http.Client{
+			Transport: tr,
+		},
+	}
+}
+
+func (c *Client) ListDevices(tokenPath string, detailLevel int) error {
+	if detailLevel > 3 || detailLevel < 0 {
+		return fmt.Errorf("FAILURE: invalid devices detail")
+	}
+	token, err := ioutil.ReadFile(tokenPath)
+	if err != nil {
+		return errors.Wrap(err, "Please Login first")
+	}
+
+	req, err := http.NewRequest(http.MethodGet, c.devicesListURL, nil)
+	if err != nil {
+		return errors.Wrap(err, "Failed to create HTTP request")
+	}
+	req.Header.Set("Authorization", "Bearer "+string(token))
+
+	reqDump, err := httputil.DumpRequest(req, false)
+	if err != nil {
+		return err
+	}
+	log.Verbf("sending request: \n%s", string(reqDump))
+
+	rsp, err := c.client.Do(req)
+	if err != nil {
+		return errors.Wrap(err, "Get /devauth/devices request failed")
+	}
+	if rsp.StatusCode != 200 {
+		return fmt.Errorf("Get /devauth/devices request failed with status %d\n", rsp.StatusCode)
+	}
+
+	defer rsp.Body.Close()
+
+	body, err := ioutil.ReadAll(rsp.Body)
+	if err != nil {
+		return err
+	}
+
+	var list devicesList
+	err = json.Unmarshal(body, &list.devices)
+	if err != nil {
+		return err
+	}
+	for _, v := range list.devices {
+		listDevice(v, detailLevel)
+	}
+
+	return nil
+}
+
+func listDevice(a deviceData, detailLevel int) {
+	fmt.Printf("ID: %s\n", a.ID)
+	fmt.Printf("Status: %s\n", a.Status)
+	if detailLevel >= 1 {
+		fmt.Println("IdentityData:")
+		if a.IdentityData.Mac != "" {
+			fmt.Printf("  Mac: %s\n", a.IdentityData.Mac)
+		}
+		if a.IdentityData.Sku != "" {
+			fmt.Printf("  Sku: %s\n", a.IdentityData.Sku)
+		}
+		if a.IdentityData.Sn != "" {
+			fmt.Printf("  SN: %s\n", a.IdentityData.Sn)
+		}
+	}
+	if detailLevel >= 1 {
+		fmt.Printf("CreatedTs: %s\n", a.CreatedTs)
+		fmt.Printf("UpdatedTs: %s\n", a.UpdatedTs)
+		fmt.Printf("Decommissioning: %t\n", a.Decommissioning)
+	}
+	if detailLevel >= 2 {
+		for i, v := range a.AuthSets {
+			fmt.Printf("AuthSet[%d]:\n", i)
+			fmt.Printf("  ID: %s\n", v.ID)
+			fmt.Printf("  PubKey:\n%s", v.PubKey)
+			fmt.Println("  IdentityData:")
+			if v.IdentityData.Mac != "" {
+				fmt.Printf("    Mac: %s\n", v.IdentityData.Mac)
+			}
+			if v.IdentityData.Sku != "" {
+				fmt.Printf("    Sku: %s\n", v.IdentityData.Sku)
+			}
+			if v.IdentityData.Sn != "" {
+				fmt.Printf("    SN: %s\n", v.IdentityData.Sn)
+			}
+			fmt.Printf("  Status: %s\n", v.Status)
+			fmt.Printf("  Ts: %s\n", v.Ts)
+		}
+	}
+
+	fmt.Println("--------------------------------------------------------------------------------")
+}
+
+func JoinURL(base, url string) string {
```

Comment:
Suggestion:
How about `path.Join` ?

Reply:
Thanks. Cleaned up also from `useradm` and `deployments`

Reply:
Nah, `path.Join` does not work nicely with URLs, really. I'll leave as is but moving it to a common place.